### PR TITLE
SALTO-1715: flatten additional properties

### PR DESCRIPTION
**Pull request by alonstern**
Added `flattenAdditionalProperties` to remove the additional properties attribute we added on fetch before deploy and used it in Jira.

Only the last commit is relevant
This PR is build over https://github.com/salto-io/salto/pull/2577

---
_Release Notes_: 
None

---
_User Notifications_: 
None

### CODE REVIEW

**Comment by netama** [comment]
@alonstern can you please link to the preceding PR explicitly so that they are linked (and easier to follow)?

File: packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  const deploymentAnnotationValue = additionalPropertiesField
+    .annotations[OPERATION_TO_ANNOTATION[action]]
+  if (deploymentAnnotationValue || deploymentAnnotationValue === undefined) {
+    _.assign(values, values.additionalProperties ?? {})
+  }
+  delete values.additionalProperties
+}
+
+/**
+ * Remove the additional properties value we added on fetch in normalizeElementValues before deploy
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts:41]: :+1: for the doc

File: packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * and adds the its values to to top level values if deployable
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts:42]: nit: `adds the` -> `add`, `to to` -> `to`

File: packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import { OPERATION_TO_ANNOTATION } from '../../../deployment/annotations'
+
+
+const removeAdditionalPropertiesFlat = async (
+  values: Values,
+  type: ObjectType,
+  action: ActionName
+): Promise<void> => {
+  const additionalPropertiesField = type.fields.additionalProperties
+  if (additionalPropertiesField === undefined) {
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts:28]: notice that there are technically cases where there is a "real" `additionalProperties` field, we may need to be more careful with them

File: packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    return
+  }
+
+  const deploymentAnnotationValue = additionalPropertiesField
+    .annotations[OPERATION_TO_ANNOTATION[action]]
+  if (deploymentAnnotationValue || deploymentAnnotationValue === undefined) {
+    _.assign(values, values.additionalProperties ?? {})
+  }
+  delete values.additionalProperties
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts:37]: what happens if we keep `additionalProperties` when it's not deployable that requires us to omit these values? won't that behave just the same as any other non-deployable field?

File: packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts (hunk 3 above)
Comment (alonstern) [reply packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts:28]: do we handle these cases when fetching the types from the swagger?

Comment (netama) [reply packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts:28]: not in an ideal way, but we do distinguish between them in a few places, e.g. [here](https://github.com/salto-io/salto/blob/main/packages/adapter-components/src/elements/swagger/type_elements/element_generator.ts#L159-L182). so technically you can differentiate based on the type (`Map<unknown>` if there wasn't such a field, and either `unknown` or another type if there wasn't). There are still some loopholes but I think if we check the type it should be good enough (unless you have a better suggestion of course :) )

File: packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  action: ActionName
+): Promise<void> => {
+  const additionalPropertiesField = type.fields.additionalProperties
+  if (additionalPropertiesField === undefined) {
+    return
+  }
+
+  const deploymentAnnotationValue = additionalPropertiesField
+    .annotations[OPERATION_TO_ANNOTATION[action]]
+  if (deploymentAnnotationValue || deploymentAnnotationValue === undefined) {
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts:34]: nit: maybe be explicit about `=== true`?

File: packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    _.assign(values, values.additionalProperties ?? {})
```
Comment (netama) [inline packages/adapter-components/src/elements/swagger/deployment/additional_properties.ts:35]: nit: maybe we should warn / fail if there are any overlaps between the `additionalProperties` keys and the existing keys? it won't happen for `additionalProperties` that we created based on the API response, but with deploy we technically also allow local edits

**Review by netama — APPROVED** [review]
I think we should still handle a few edge cases, pre-approving for once these are handled

**Pull request merged**